File: coding/src/com/algos/li21_design/InsertDeleteGetrandomO1.java
package  com.algos.li21_design;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Random;

public class InsertDeleteGetrandomO1 {

	public static void main(String[] args) {
		InsertDeleteGetrandomO1 obj = new InsertDeleteGetrandomO1();
		
		System.out.println(obj.insert(1));
		System.out.println(obj.remove(2));
		System.out.println(obj.insert(2));
		System.out.println(obj.getRandom());
		System.out.println(obj.remove(1));
		System.out.println(obj.insert(2));
		System.out.println(obj.getRandom());
	}

	/*
	 * Problem: https://leetcode.com/problems/insert-delete-getrandom-o1/
	 * Companies: Amazon
	 * */
	HashMap<Integer, Integer> map;
    List<Integer> list;
    Random r;
    public InsertDeleteGetrandomO1() {
        map = new HashMap<>();
        list = new ArrayList<>();
        r = new Random();
    }
    
    public boolean insert(int val) {
        if(!map.containsKey(val)){
            map.put(val, list.size());
            list.add(val);
            
            return true;
        }
        
        return false;
    }
    
    public boolean remove(int val) {
        if(map.containsKey(val)){
            int valInd = map.get(val);
            int lastOne = list.get(list.size()-1);
            
            if(valInd < list.size()-1){
                list.set(valInd, lastOne);
                map.put(lastOne, valInd);                
            }
            
            list.remove(list.size()-1);
            map.remove(val);
            
            return true;
        }
        
        return false;
    }
    
    public int getRandom() {
        int ind = r.nextInt(list.size());
        return list.get(ind);
    }
}

File: dsa/topics/16_design/python/insert_delete_getrandom_o1.py
import random

# Problem: https://leetcode.com/problems/insert-delete-getrandom-o1/
# Companies: Amazon
class InsertDeleteGetRandomO1:

    def __init__(self):
        self.map = {}          # val -> index in list
        self.list = []         # stores values
        self.rand = random.Random()

    # Idea:
    # If val is not present, add it to the end of list
    # Store its index in map
    # Time: O(1)
    def insert(self, val: int) -> bool:
        if val not in self.map:
            self.map[val] = len(self.list)
            self.list.append(val)
            return True
        return False

    # Idea:
    # To delete in O(1), swap the element to remove with the last element
    # Update index of swapped element in map
    # Remove last element from list and delete from map
    # Time: O(1)
    def remove(self, val: int) -> bool:
        if val in self.map:
            idx = self.map[val]
            last_val = self.list[-1]

            if idx < len(self.list) - 1:
                self.list[idx] = last_val
                self.map[last_val] = idx

            self.list.pop()
            del self.map[val]
            return True
        return False

    # Idea:
    # Pick a random index from list
    # Time: O(1)
    def getRandom(self) -> int:
        return self.list[self.rand.randint(0, len(self.list) - 1)]


# ------------------ Test (Same as Java main) ------------------

obj = InsertDeleteGetRandomO1()
print(obj.insert(1))     # True
print(obj.remove(2))     # False
print(obj.insert(2))     # True
print(obj.getRandom())   # 1 or 2
print(obj.remove(1))     # True
print(obj.insert(2))     # False
print(obj.getRandom())   # 2
